beluga_ros/tools/conversion_utils.py

```diff
@@ -0,0 +1,273 @@
+            total_cells = covs.shape[0]
+
+            assert covs.shape[1] == 2
+            assert covs.shape[2] == 2
+            assert means.shape[1] == 2
+            assert discrete_cells.shape[0] == total_cells
+            assert discrete_cells.shape[1] == 2
+            assert means.shape[0] == total_cells
+
+            ret = NDTMap(resolution=resolution)
+            for cell_num in range(total_cells):
+                cell = DiscreteCell(*discrete_cells[cell_num].tolist())
+                mean = means[cell_num]
+                cov = covs[cell_num]
+                ret.add_distribution(
+                    cell=cell, ndt=NormalDistribution(mean=mean, covariance=cov)
+                )
+            return ret
+
+
+@dataclass
+class OccupancyGrid:
+    resolution: float
+    origin: np.ndarray
+    grid: np.ndarray
+
+    @staticmethod
+    def load_from_file(yaml_path: Path) -> "OccupancyGrid":
+        """Create an grid from a yaml file describing a ROS style occupancy grid."""
+        assert yaml_path.is_file(), f"file {yaml_path} doesn't exist."
+        assert yaml_path.suffix == ".yaml"
+
+        with open(yaml_path, "rb") as fp:
+            data = yaml.safe_load(fp)
+
+        pgm_path: Path = yaml_path.parent / data["image"]
+        assert pgm_path.is_file(), f".pgm file at {pgm_path} does not exist"
+
+        with open(yaml_path.parent / data["image"], "rb") as pgmf:
+            grid = plt.imread(pgmf)
+
+        return OccupancyGrid(
+            resolution=data["resolution"], origin=np.asarray(data["origin"]), grid=grid
+        )
+
+
+def grid_to_point_cloud(occupancy_grid: OccupancyGrid) -> np.ndarray:
+    """
+    Convert an OccupancyGrid's occupied cells to a 2D point cloud.
+
+    Uses the center of the cell for the conversion to reduce max error.
+    """
+    occupied_cells_indices = np.asarray(
+        np.where(occupancy_grid.grid == 0)
+    )  # in ROS maps 0 == occupied.
+    res = occupancy_grid.resolution
+
+    # Discretized occupied cells using the center of the cell.
+    points = occupied_cells_indices * res + np.ones_like(occupied_cells_indices) * (
+        res / 2
+    )
+    # Compensate for origin
+    points[0] += occupancy_grid.origin[0]
+    points[1] += occupancy_grid.origin[1]
+    return points
+
+
+def fit_normal_distribution(points: np.ndarray) -> Optional[NormalDistribution]:
+    """Fit a normal distribution to a set of 2D points."""
+    assert points.shape[0] == 2
+    # Literature suggests doing this check to avoid singularities.
+    # See The Three-Dimensional Normal-Distributions Transform– an Efficient
+    # Representation for Registration, Surface Analysis, and Loop Detection
+    # by Martin Magnusson, 2009 chapter 6.
+    if points.shape[1] <= 5:
+        return None
+    mean = points.mean(axis=1)
+    covariance = np.cov(points)
+
+    # avoid singularities by enforcing a minimum variance in both dimensions.
+    covariance[0, 0] = max(covariance[0, 0], 5e-3)
```

Comment:
@serraramiro1 nit: can we make the minimum variance a parameter?

Reply:
Done